sections/generic-best-practices/test-code-with-debugging-in-mind.md

```diff
@@ -0,0 +1,345 @@
+};
+```
+
+I'm not too fond of the above example because
+
+1.  It tries to cover too many use cases at a time.
+2.  If it fails, you have to deal with conditions that make the whole experience a nightmare.
+
+You can find more best practices we follow internally in the [Hasura Console UI coding patterns: Testing](https://dev.to/noriste/hasura-console-ui-coding-patterns-testing-281d) article.
+
+### Matching the test's code and test runner's commands
+
+
+![The test code side by side with the Cypress panel with some red arrows to match the code and the Cypress logs](../../assets/images/test-code-with-debugging-in-mind/no-match-between-code-and-runner.jpg)
+
+
+
+
+The Cypress Test Runner helps understand what is happening in the application and which commands are executed, but when you debug a test, it is hard to have an immediate correlation between the Test Runner and the code. More, the logs do not help in understanding what the test is doing in feature terms (ex., the logs say "types in the textarea" but do not say "Type in the Type Configuration textarea"). So, detecting the root of a failure is hard. Cypress records videos for the failing tests, but it is useless if the reader/debugger does not have an immediate correlation between the logs and what the test is doing in plain English.
+
+#### Look at something like this
+
+
+![The code and the Cypress panel side by side, with a lot of custom logs that allow directly connecting what happens in Cypress to a precise point in the code](../../assets/images/test-code-with-debugging-in-mind/match-between-code-and-runner.jpg)
+
+
+
+I add a log reporting what the test is doing, allowing a direct correlation between the test's code and the Test Runner (`cy.log('**--- Type in the Webhook Handler field**');`).
+
+Please note that you can pass more arguments to 'cy.log', and they are logged right in the devtools' console when you click on the logged command.
+
+
+![The Cypress Test Runner showing the value of a logged object](../../assets/images/test-code-with-debugging-in-mind/cy-log-console.jpg)
+
+
+
+Storybook and Playwright already have the concept of `step` utilities that allow explaining in English what is happening in the test. Cypress does not have the same option, so the `cy.log` I propose is valuable, in my opinion.
+
+### Use clear selectors
+
+Look at this code
+
+```ts
+cy.get('textarea')
+  .eq(0)
+  .type(`{enter}{uparrow}${statements.createMutationGQLQuery}`, {
+    force: true,
+  });
+```
+
+What is `cy.get('textarea').eq(0)`? In the absence of better selectors, I suggest hiding them under Cypress aliases, such as
+
+```ts
+// Assign an alias to the most unclear selectors for future references
+cy.get('textarea').eq(0).as('actionDefinitionTextarea');
+cy.get('textarea').eq(1).as('typeConfigurationTextarea');
+```
+
+and then referring to them this way
+
+```ts
+cy.get('@actionDefinitionTextarea').clearConsoleTextarea().type(/* ... */);
+```
```

Comment:
or we can make a function

const actionDefinitionTextarea = () => cy.get('textarea').eq(0)

Reply:
I'll leave it to you. Imo aliases are weak because they are scoped to the it block under test. I wish this plugin worked https://github.com/bahmutov/cypress-aliases and it was the default for all of Cypress. Then we would be talking!